### DISCUSSION TITLE
Generate MyPy types from SDL

### DISCUSSION THREAD

**Discussion by bartenra** [mirumee/ariadne#509] [Q&A]
Hi there.

One of the great things about using GraphQL with Typescript is that you can generate types for everything from schemas, to enums, to resolvers, etc. etc.

Is it also possible to add generated MyPy type hints to Ariadne resolvers? I couldn't find anything about this in the docs.

**Comment by rafalp** [comment]
Can you show typescript examples of generated resolvers? I was under impression that Apollo's Codegen is only for the client side. 🤔

**Reply by bartenra** [reply]
https://www.graphql-code-generator.com/docs/plugins/typescript-resolvers

https://the-guild.dev/blog/better-type-safety-for-resolvers-with-graphql-codegen

It's a project called GraphQL Codegen that supports it.
If my team decides to use Ariadne today, I might give it a go to implement a Ariadne plugin for it.
It has a very extensible plugin architecture. :-)
https://www.graphql-code-generator.com/docs/custom-codegen/using-visitor

Should be possible to hack something together with `Literal` and `@overload`.

**Reply by bartenra** [reply]
Here's the demo from that blog post I shared: https://codesandbox.io/s/condescending-albattani-ixyp9?file=/resolvers.ts

Maybe something like this?

SDL:

```graphql
type Query {
  hello: String!
}
```

Generated:

```py
class ResolversProtocol(Protocol):
    @staticmethod
    def hello_resolver(_: None, info: GraphQLResolveInfo) -> str: ...
```

Implementation:

```py
class Resolvers(ResolversProtocol):
    @staticmethod
    def hello_resolver(_: None, info: GraphQLResolveInfo) -> str:
        return "hello world"
```

It's a bit silly that you have to repeat the types of a Protocol with `mypy`.

**Comment by bartenra** [comment]
@rafalp I would like to invite you to take a look at a repo with a proof of concept: :+1: 

https://github.com/enra-GmbH/graphql-codegen-ariadne

**Comment by fernandogrd** [comment]
Just found out that someone started a pure python graphql code generator:
https://github.com/sauldom102/gql_schema_codegen

**Comment by tianhuil** [comment]
Any progress on this @rafalp: having typings on both the client and the server is key -- otherwise, we have much weaker guarantees that the client will be able to talk to the server.

**Reply by rafalp** [reply]
Nobody picked this to my knowledge, so no, no progress.

Personally I would like to first see a proposal of generated types for [challenge schema](https://github.com/mirumee/ariadne-codegen/issues/122#issuecomment-1499207518) I've posted. Lot of folk keep bringing up an idea that there should be MyPy types for resolvers, but those proposals skim the fact that `User` in schema is usually completely different from `UserModel` in Python, and there's resolvers layer on `User` fields than map from one to the other.
